packages/runtime/src/runtimes/react/__tests__/react-runtime.test.tsx

```diff
@@ -0,0 +1,104 @@
+/** @jest-environment jsdom */
+import { type MouseEvent } from 'react'
+import {
+  Checkbox,
+  Combobox,
+  Color,
+  Image,
+  Link,
+  List,
+  Number,
+  Select,
+  Shape,
+  TextArea,
+  TextInput,
+} from '../../../controls'
+
+import { ReactRuntime } from '../react-runtime'
+
+type Card = {
+  imageSrc?: string
+  imageAlt: string
+  title: string
+  text: string
+  link: {
+    href: string
+    onClick: (event: MouseEvent) => void
+  }
+}
+
+type Entity = {
+  id: number
+  name: string
+}
+
+function Sandbox(props: {
+  cards: Card[]
+  entityId?: Entity['id']
+  theme: 'light' | 'dark'
+  background: string
+  boolean: boolean
+  number?: number
+}) {
+  return <div>{JSON.stringify(props)}</div>
+}
+
+describe('registerComponent', () => {
+  test("correctly deduces control definitions' resolved value types", () => {
+    ReactRuntime.registerComponent(Sandbox, {
```

Comment:
Reproduce for the issue, prior to the fix, this would fail with

<img width="1078" alt="repro" src="https://github.com/user-attachments/assets/d9d704fa-7b19-4c73-95d3-ed77d6c15fa4">

Reply:
I ran into this while upgrading our examples to 0.21.0. We tested 0.20.0 with both makeswift.com and bigcommerce.com implementations, but not with our own examples.

Reply:
Do we know what about this particular registration was resulting in the errors? I assume it was the reference to a subtype.

Reply:
The list of `Combobox` options was deduced as `{ id: string; value: Data; label: string }` instead of expected `{ id: string; value: number; label: string }`. If you factored out the `Combobox` prop def into a standalone variable, though, all of the types would be deduced properly:

```typescript
const entityId = Combobox({
  async getOptions() {
    return fetch(`/api/entities`)
      .then(r => r.json())
      .then((entities: Entity[]) =>
        entities.map(entity => ({
          id: entity.id.toString(),
          label: entity.name,
          value: entity.id,
        })),
      )
  },
  label: 'Entity',
})

...

ReactRuntime.registerComponent(Sandbox, {
   type: 'sandbox',
      label: 'Sandbox',
      props: {
         ...
         entityId, // this is fine
     ...
```

So the deduction was only failing in the context of `registerComponent`.

As far as what is specific about `Combobox` vs other control definitions, my guess would be is that `Combobox` is the only control which resolved value type is derived from the return type of a function.